sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -4812,47 +4918,68 @@ piEnqueueKernelLaunch(pi_queue Queue, pi_kernel Kernel, pi_uint32 WorkDim,
   // Lock automatically releases when this goes out of scope.
   std::lock_guard<std::mutex> QueueLock(Queue->PiQueueMutex);
 
-  _pi_ze_event_list_t TmpWaitList;
+  bool UseCopyEngine = false;
 
+  Queue->EventlessMode |= !Event;
+  if (Queue->EventlessMode) {
+    if (auto Res = Queue->AddBarrierInPreviousCmdListIfNeeded(UseCopyEngine))
+      return Res;
+  }
+
+  _pi_ze_event_list_t TmpWaitList;
   if (auto Res = TmpWaitList.createAndRetainPiZeEventList(NumEventsInWaitList,
                                                           EventWaitList, Queue))
     return Res;
 
   // Get a new command list to be used on this call
   pi_command_list_ptr_t CommandList{};
   if (auto Res = Queue->Context->getAvailableCommandList(
-          Queue, CommandList, false /* UseCopyEngine */,
-          true /* AllowBatching */))
+          Queue, CommandList, UseCopyEngine, true /* AllowBatching */))
     return Res;
+  Queue->LastCommandList = CommandList;
 
   ze_event_handle_t ZeEvent = nullptr;
-  pi_result Res = createEventAndAssociateQueue(
-      Queue, Event, PI_COMMAND_TYPE_NDRANGE_KERNEL, CommandList);
-  if (Res != PI_SUCCESS)
-    return Res;
-  ZeEvent = (*Event)->ZeEvent;
-  (*Event)->WaitList = TmpWaitList;
+  if (Event) {
+    pi_result Res = createEventAndAssociateQueue(
+        Queue, Event, PI_COMMAND_TYPE_NDRANGE_KERNEL, CommandList);
+    if (Res != PI_SUCCESS)
+      return Res;
+    ZeEvent = (*Event)->ZeEvent;
+    (*Event)->WaitList = TmpWaitList;
 
-  // Save the kernel in the event, so that when the event is signalled
-  // the code can do a piKernelRelease on this kernel.
-  (*Event)->CommandData = (void *)Kernel;
+    // Save the kernel in the event, so that when the event is signalled
+    // the code can do a piKernelRelease on this kernel.
+    (*Event)->CommandData = (void *)Kernel;
+  } else {
+    auto &CommandListInfo = CommandList->second;
+    ++CommandListInfo.NumEventlessCommands;
+    if (TmpWaitList.Length != 0) {
+      CommandListInfo.PiEventLists.emplace_back(TmpWaitList.PiEventList);
+      CommandListInfo.ZeEventLists.emplace_back(TmpWaitList.ZeEventList);
+      CommandListInfo.Lengths.emplace_back(TmpWaitList.Length);
+    }
+    CommandListInfo.EventlessKernelsInUse.emplace_back(Kernel);
+  }
```

Comment:
What is the reason to duplicate data structures into a completely different set of logic? I think we should rather continue to create PI Event for bookkeeping purposes (but not create L0 event along). That Pi event wouldn't be sent back to the user but kept in the existing PI command-list structure.

Reply:
Thanks, I addressed your comment.